libraries/AP_Math/fourier.cpp

```diff
@@ -0,0 +1,238 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+ 
+ //designed by Philippe Crochat / Anemos Technologies : pcrochat@anemos-technologies.com
+
+#include "fourier.h"
+
+Fourier_Analysis::Fourier_Analysis(){
+	int i;
+	
+	for(i=0;i<2;i++)
+	{
+		_fourier_transform[i].x=0.0f;
+		_fourier_transform[i].y=0.0f;
+	}
+
+	_time_elapsed=0.0f;
+	_buffer_index=0;
+	
+	_signal_mean.x=0.0f;
+	_signal_mean.y=0.0f;
+	_signal_mean.z=0.0f;
+	
+	_angle_tot.x=0.0f;
+	_angle_tot.y=0.0f;
+	
+	_result_filter.set_cutoff_frequency(RESULT_FILTER_FREQ);
+	
+	_last_result.x=0.0f;
+	_last_result.y=0.0f;
+}
+
+void Fourier_Analysis::Set_Fourier_Analysis(int buffer_size){
+	int i;
+
+	_buffer_size=buffer_size;
+	
+	_signal=(Vector3f *)malloc(buffer_size*sizeof(Vector3f));
+	_timing=(Timing_Struct *)malloc(buffer_size*sizeof(Timing_Struct));
+	_phase=(float *)malloc(buffer_size*sizeof(float));
+	
+	for(i=0;i<_buffer_size;i++)
+	{
+		_signal[i].x=0.0f;
+		_signal[i].y=0.0f;
+		_signal[i].z=0.0f;
+		
+		_timing[i].dt=0.0f;
+		
+		_timing[i].omega.x=0.0f;
+		_timing[i].omega.y=0.0f;
+		_timing[i].omega.z=0.0f;
+		
+		_timing[i].T=0.0f;
+		
+		_phase[i]=0.0f;
+	}
+}
+
+Fourier_Analysis::~Fourier_Analysis(){
+	free(_signal);
+	free(_timing);
+	free(_phase);
+}
+
+void Fourier_Analysis::accumulate(Vector3f *new_sample, Timing_Struct *new_timing){
+	int previous_index;
+	int i;
+	float actual_heading;
+	float actual_yaw;
+	
+	previous_index=_buffer_index-1;
+
+	if(previous_index<0)
+		previous_index=_buffer_size-1;
+	
+	Vector3f signal_current=_signal[_buffer_index];//-_signal_mean/_buffer_size;
+	Vector2f fourier_coef;
+	
+	Vector2f new_result;
+	
+	fourier_coef.x=cosf(_phase[_buffer_index]);
+	fourier_coef.y=-sinf(_phase[_buffer_index]);
+	
+	for(i=0;i<2;i++)
+	{
+		_fourier_transform[i].x -= (fourier_coef.x*signal_current.x);
+		_fourier_transform[i].y -= (fourier_coef.y*signal_current.x);
```

Comment:
Current y?

Reply:
nope the sin and cos coefficient apply to the same signal but there is a bug with the loop, when i==1 the fourier coef should be applied to the signal_current.y